### DISCUSSION TITLE
Cannot use import/export statements outside a module

### DISCUSSION THREAD

**Discussion by flash168** [sebastienros/jint#1472] [Q&A]
Jint.Runtime.JavaScriptException:“Cannot use import/export statements outside a module”
js脚本内使用了import aa from"https://ww.aaa.com/a.js";

**Comment by lahma** [comment]
I think that follows the spec, imports are valid inside module

**Comment by snuup** [comment]
but how can we then use a javascript module inside the "main" code that we run inside Evalute?

**Reply by lahma** [reply]
`Evaluate()`/`Execute()` run classic **scripts**, and per the ECMAScript spec `import`/`export` declarations only exist inside **modules** — that part can't change. So you either make your "main" code a module too, or hand the module's exports to the script another way. Three options, pick whichever fits:

**Option 1 (recommended): run your main code as a module**

```csharp
var engine = new Engine();
engine.Modules.Add("lib", "export function add(a, b) { return a + b; }");
engine.Modules.Add("main", """
    import { add } from 'lib';
    globalThis.result = add(1, 2);
    """);
engine.Modules.Import("main"); // runs main, imports resolve normally
var result = engine.Evaluate("result"); // 3
```

For modules on disk, use `new Engine(options => options.EnableModules(@"C:\Scripts"))` and `engine.Modules.Import("./main.js")` — relative imports between files then work as expected. Note there's no completion value for modules; export what you need, or stash it on `globalThis` as above.

**Option 2: import from .NET, expose the namespace as a global, keep using `Evaluate`**

```csharp
var engine = new Engine();
engine.Modules.Add("lib", "export function add(a, b) { return a + b; }");
var ns = engine.Modules.Import("lib");  // returns the module namespace object
engine.SetValue("lib", ns);
var result = engine.Evaluate("lib.add(20, 22)"); // 42
```

**Option 3: dynamic `import()` inside the script**

The spec does allow the dynamic `import()` *expression* in scripts, and Jint supports it — it returns a promise:

```csharp
var engine = new Engine();
engine.Modules.Add("lib", "export function add(a, b) { return a + b; }");
engine.Evaluate("import('lib').then(m => { globalThis.result = m.add(40, 2); });");
var result = engine.Evaluate("result"); // 42
```

Modules can also be defined programmatically from .NET (`engine.Modules.Add("lib", builder => builder.ExportValue(...).ExportFunction(...).ExportType<T>())`) and then imported from JS the same way. One thing to keep in mind: module code always runs in strict mode, so sloppy-mode patterns from script code won't carry over.
